File: uva10327/uva10327.cpp
#include <iostream>
#include <algorithm>
using namespace std;
int arr[10000];
int tmp[10000];


int solve_aux(int low,int mid,int hi){
	int i = low,j = mid,ans = 0,k = 0;
	while(i<mid && j < hi){
		if(arr[i]<=arr[j]){
			tmp[k] = arr[i];
			i++;
		}
		else{
			tmp[k] = arr[j];
			ans += mid-i;
			j++;
		}
		k++;
	}
	while(i < mid){
		tmp[k] = arr[i];
		i++;
		k++;
	}
	while(j < hi){
		tmp[k] = arr[j];
		j++;
		k++;
	}
	k = 0;
	for(i = low; i < hi;i++){
		arr[i] = tmp[k];
		k++;
	}
	return ans;

}

int solve(int low,int hi){
	int ans,mid;
	if (low == hi || low +1 == hi)
		ans = 0;
	else{
		mid = (low+hi)/2;
		ans = solve(low,mid);
		ans += solve(mid,hi);
		ans += solve_aux(low,mid,hi);
	}
	return ans;
}


int main(){
	int n,v,i;
	while(cin>>n){
		for(i = 0; i<n;i++){
			cin>>v;
			arr[i] = v;
		}
		cout<<"Minimum exchange operations : "<<solve(0,n)<<endl;
	}
}

File: uva10327/uva10327.py
from sys import stdin
import sys
sys.setrecursionlimit(10000000)

def solve_aux(arr,low,mid,hi):
    i,j,ans = low,mid,0
    tmp = []
    while i < mid and j < hi:
        if arr[i] <= arr[j]:
            tmp.append(arr[i])
            i += 1
        else:
            tmp.append(arr[j])
            j += 1
            ans += mid-i
    while i < mid:
        tmp.append(arr[i])
        i += 1
    while j < hi:
        tmp.append(arr[j])
        j += 1
    arr[low:hi] = tmp
    return ans

def solve(arr,low,hi):
    ans = None
    if low == hi or low+1 == hi:
        ans = 0
    else:
        mid = (low+hi)//2
        ans = solve(arr,low,mid)
        ans += solve(arr,mid,hi)
        ans += solve_aux(arr,low,mid,hi)
    return ans


def main():
    n = stdin.readline()
    while n != "" and n != "\n" and len(n):
        n = int(n)
        arr = [int(x) for x in stdin.readline().split()]
        #print(arr)
        print("Minimum exchange operations : {0}".format(solve(arr,0,n)))
        n = stdin.readline()
main()
